infra/baseServer.go

```diff
@@ -358,47 +368,56 @@ func (bs *BaseServer[T]) initialiseAgents(m []AgentGeneratorCountPair[T]) {
 }
 
 // generate a server instance based on a mapping function and number of iterations
-func CreateServer[T IAgent[T]](generatorArray []AgentGeneratorCountPair[T], iterations int, maxDuration time.Duration, agentServerChannelBufferSize int) *BaseServer[T] {
+func CreateServer[T IAgent[T]](generatorArray []AgentGeneratorCountPair[T], iterations int, maxDuration time.Duration,agentAgentChanBuffer,serverAgentChanBuffer int) *BaseServer[T] {
 	serv := &BaseServer[T]{
 		agentMap:               make(map[uuid.UUID]T),
 		agentIdSet:             make(map[uuid.UUID]struct{}),
-		agentAgentChannelMap:   make(map[uuid.UUID]chan IMessage),
+		agentAgentChannelMap:   make(map[uuid.UUID]chan IMessage[T]),
 		serverAgentChannelMap:  make(map[uuid.UUID]chan ServerNotification),
 		closureChannel:         make(chan uuid.UUID),
 		waitEnd:                &sync.WaitGroup{},
 		listeningWaitGroup:     &sync.WaitGroup{},
 		agentStoppedTalkingMap: make(map[uuid.UUID]struct{}),
-		agentServerChannel:     make(chan uuid.UUID, agentServerChannelBufferSize),
+		agentServerChannel:     nil,
 		maxMessagingDuration:   maxDuration,
 		roundRunner:            nil,
 		iterations:             iterations,
+		agentAgentChanBufferSize: agentAgentChanBuffer,
+		serverAgentChanBufferSize: serverAgentChanBuffer,
 	}
 	serv.initialiseAgents(generatorArray)
+	serv.agentServerChannel = make(chan uuid.UUID, len(serv.agentMap))
 	return serv
 }
 
-func listenOnChannel[T IAgent[T]](a T, agentAgentchannel chan IMessage, serverAgentchannel chan ServerNotification, wait *sync.WaitGroup) {
+func listenOnChannel[T IAgent[T]](a T, agentAgentchannel chan IMessage[T], serverAgentchannel chan ServerNotification, wait *sync.WaitGroup) {
 	defer wait.Done()
 
 	// checkMessageHandler()
 
 	listenAgentChannel := false
-	fmt.Println("started listening", a.GetID())
+	//fmt.Println("started listening", a.GetID())
 
 listening:
 	for {
+		if a.GetListeningSpinnerFlag() {
+			fmt.Println("STOPPED SPINNER", a.GetID())
+			a.SetListeningSpinner(false)
```

Comment:
I use a flag so that the agent can break out of the listening spinner whenever it wants. If this poses a synchronisation problem then we can always get it to send a request to the server to allow it to exit but I cannot see why it would pose a problem.